### PR TITLE
Review Class Implementation 

### PR DESCRIPTION
**Pull request by ghost**
@all
Please review the class-based implementation of the code.

The code fetches the lyrics of any song based on user input. It first searches the database for a possible match. If a match is found, it returns the match.

If a match is not found, it requests the Fetch API to get a response, then filters the response and saves the new data in the database. It then shows the results to the user.

Please review the code and suggest any necessary changes.

[Screencast from 19-06-24 12:04:55 PM IST.webm](https://github.com/mohit-trootech/Music-Lyrics/assets/172255945/b70e1460-bb03-44cb-9021-ebc2cad7fdd2)

### CODE REVIEW

File: lyricx_database.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    cursor = db.cursor()
+    cursor.execute("CREATE TABLE IF NOT EXISTS "
+                   "tracks (id, track_name, album_name, artist_name, rating, track_lyrics);")
+
+    @classmethod
+    def fetch_from_database(cls, song):
+        cls.cursor.execute(f"SELECT * FROM tracks where track_name = '{song}';")
+        return cls.cursor.fetchone()
+
+    @classmethod
```
Comment (yatishTrootech) [inline lyricx_database.py]: add docstring in every method and type hinting and return type

File: lyricx_database.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import sqlite3
+
+
+class Database:
+    db = sqlite3.connect("music.db")
+    cursor = db.cursor()
+    cursor.execute("CREATE TABLE IF NOT EXISTS "
+                   "tracks (id, track_name, album_name, artist_name, rating, track_lyrics);")
+
+    @classmethod
```
Comment (yatishTrootech) [inline lyricx_database.py]: create class 
```
Request:
     
     def parse_data(response):
           if res.coede != 200 then raise
           return response.json()

     def get(url):
           res = request.get(url)
           return self.parse_data(res)
```

**Review by yatishTrootech — CHANGES_REQUESTED** [review]

File: lyricx_database.py (hunk 1 above)
Comment (ghost) [reply lyricx_database.py]: Ok, Will Implement on Every Next Project

File: lyricx_database.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    def fetch_from_database(cls, song):
+        cls.cursor.execute(f"SELECT * FROM tracks where track_name = '{song}';")
```
Comment (yatishTrootech) [inline lyricx_database.py]: i think it should be case insensitive and like query

**Review by yatishTrootech — CHANGES_REQUESTED** [review]

File: lyricx_database.py (hunk 3 above)
Comment (ghost) [reply lyricx_database.py]: No, = should be there since many songs with same keyword that's why i used equals to.

File: lyricx.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+from requests import get
+
+import constant
+import lyricx_database
+
+FETCH_TRACK_URL = (f'https://api.musixmatch.com/ws/1.1/track.search?apikey={constant.API_KEY}&s_track_rating=DESC'
+                   '&s_has_lyrics&page_size=1&page=1&q_track=')
+FETCH_TRACK_LYRICS = (f'https://api.musixmatch.com/ws/1.1/track.lyrics.get?apikey={constant.API_KEY}'
+                      f'&track_id=')
```
Comment (Dhruvi-Makwana) [inline lyricx.py]: If this is constant variable then move this into constant.py file.
also add error message or success message to this file.

File: lyricx.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    @staticmethod
+    def parse_data(response):
+        if response['header']['status_code'] != 200:
+            raise Exception("Boo! HTTP Error")
+        return response['body']
+
+    @classmethod
+    def get(cls, url):
+        response = requests.get(url)
+        return Request.parse_data(response.json()['message'])
```
Comment (yatishTrootech) [inline lyricx.py]: it should be instance method and self.parse_data

File: lyricx.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+                      f'&track_id=')
+
+
+class Request:
+    """This is a Python class with the purpose of collecting and parsing data. It seems like the class is supposed to
+    fetch the data from a server and check the status code that comes back. If the status code is not 200,
+    then an error is raised."""
+    @staticmethod
+    def parse_data(response):
+        if response['header']['status_code'] != 200:
```
Comment (yatishTrootech) [inline lyricx.py]: handle try except

Comment (yatishTrootech) [inline lyricx.py]: STATUS_OK=200 in constants.py and use from there

File: lyricx.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+            raise Exception("Boo! HTTP Error")
```
Comment (yatishTrootech) [inline lyricx.py]: exception msg should be from constants.py

File: lyricx.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    are located in the database, the code delivers the result to exhibit the lyrics. If the lyrics are absent from
+    the database, the code will resort to the Request class to obtain the data from the API, subsequently storing the
+    result in the database, and ultimately displaying the result.ves the result into database and prints the result"""
+
+    track_details = {}
+
+    def __init__(self, song):
+        self.song = song.capitalize().strip()
+
+    def fetch_lyrics(self):
```
Comment (yatishTrootech) [inline lyricx.py]:
```
def fetch_lyrics(self): -> None

```

File: lyricx.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+                Lyrics.get_lyrics(track_details)
+            else:
+                print(f"Not Found in Database")
+                Lyrics.request_track(self.song)
+        except sqlite3.OperationalError as e:
+            print(f"Data Not Found {e}")
+            Lyrics.request_track(self.song)
+
+    @staticmethod
+    def request_track(song):
```
Comment (yatishTrootech) [inline lyricx.py]: docstring in every method

File: lyricx.py [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+        response = Request.get(f"{FETCH_TRACK_LYRICS}{track_id}")
+        try:
+            lyrics = response['lyrics']['lyrics_body']
+            Lyrics.track_details.setdefault("lyrics", lyrics)
+            Lyrics.save_lyrics(Lyrics.track_details)
+        except KeyError:
+            print("Key Error Response Object keys Not Matched")
+
+    @staticmethod
+    def save_lyrics(track_details):
```
Comment (yatishTrootech) [inline lyricx.py]:
```
def save_lyrics(track_details: list) -> None:
```

File: lyricx.py [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+class Lyrics:
+    """The code provided is an illustration of a Python class used for acquiring lyrics for a song. When the lyrics
+    are located in the database, the code delivers the result to exhibit the lyrics. If the lyrics are absent from
+    the database, the code will resort to the Request class to obtain the data from the API, subsequently storing the
+    result in the database, and ultimately displaying the result.ves the result into database and prints the result"""
+
+    track_details = {}
+
+    def __init__(self, song):
+        self.song = song.capitalize().strip()
```
Comment (yatishTrootech) [inline lyricx.py:36]:
```
self.request = Request()
```

File: lyricx.py [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+                print(f"Not Found in Database")
+                Lyrics.request_track(self.song)
+        except sqlite3.OperationalError as e:
+            print(f"Data Not Found {e}")
+            Lyrics.request_track(self.song)
+
+    @staticmethod
+    def request_track(song):
+        try:
+            track_response = Request.get(f"{FETCH_TRACK_URL}{song}")
```
Comment (yatishTrootech) [inline lyricx.py]:
```
self.request.get
```

File: lyricx.py [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+            track_details = lyricx_database.Database.fetch_from_database(self.song)
+            if track_details:
+                print("Found in Database")
+                Lyrics.get_lyrics(track_details)
+            else:
+                print(f"Not Found in Database")
+                Lyrics.request_track(self.song)
+        except sqlite3.OperationalError as e:
+            print(f"Data Not Found {e}")
+            Lyrics.request_track(self.song)
```
Comment (yatishTrootech) [inline lyricx.py]:
```
self.request_track
```

Instance or static methiod should be called from self

File: lyricx.py [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+                "album": tracks['album_name'],
+                "artist": tracks['artist_name'],
+                "rating": tracks["track_rating"]
+            }
+            Lyrics.request_lyrics(Lyrics.track_details['id'])
+        except KeyError:
+            print("Key Error Response Object keys Not Matched")
+
+    @staticmethod
+    def request_lyrics(track_id):
```
Comment (yatishTrootech) [inline lyricx.py]:
```
self.save_lyrics
```

`request_lyrics` this should not be static method

**Review by yatishTrootech — CHANGES_REQUESTED** [review]

File: lyricx.py [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+            self.track_details.setdefault("lyrics", lyrics)
+            self.save_lyrics()
+        except KeyError as e:
+            print(constant.KEY_ERROR, e)
+
+    def save_lyrics(self) -> None:
+        """method to store fetched track details and lyrics"""
+        try:
+            Database.save_2_database(self.track_details)
+            self.get_lyrics(list(self.track_details.values()))
```
Comment (yatishTrootech) [inline lyricx.py]: no need to convert into list again

File: lyricx_database.py [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+import sqlite3
+import constant
+
+
+class Database:
+    """This class will do just that. It can fetch and save data, giving you more options for how you work with your
+    data. It’s easy to use and implement, so you can have a functional database connection in no time. sql queries to
+    fetch and save data in database"""
+
+    db = sqlite3.connect("music.db")
```
Comment (yatishTrootech) [inline lyricx_database.py]: init

Comment (yatishTrootech) [inline lyricx_database.py]: db name dynamic

**Review by yatishTrootech — CHANGES_REQUESTED** [review]

**Review by yatishTrootech — APPROVED** [review]

**Review by yatishTrootech — APPROVED** [review]

**Review by Dhruvi-Makwana — APPROVED** [review]

**Pull request merged**